sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/utils/XdsRouterUtil.java

```diff
@@ -0,0 +1,100 @@
+import io.sermant.core.service.ServiceManager;
+import io.sermant.core.service.xds.XdsCoreService;
+import io.sermant.core.service.xds.XdsServiceDiscovery;
+import io.sermant.core.service.xds.entity.ServiceInstance;
+import io.sermant.core.service.xds.entity.XdsLocality;
+import io.sermant.core.utils.NetworkUtils;
+import io.sermant.core.utils.StringUtils;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * XdsRouterUtil
+ *
+ * @author daizhenyu
+ * @since 2024-08-29
+ **/
+public class XdsRouterUtil {
+    private static XdsServiceDiscovery serviceDiscovery = ServiceManager.getService(XdsCoreService.class)
+            .getXdsServiceDiscovery();
+
+    private static XdsLocality xdsLocality;
+
+    private XdsRouterUtil() {
+    }
+
+    /**
+     * get XdsLocality of self-service
+     *
+     * @return XdsLocality
+     */
+    public static Optional<XdsLocality> getLocalityInfoOfSelfService() {
+        if (xdsLocality != null) {
+            return Optional.of(xdsLocality);
+        }
+        synchronized (XdsRouterUtil.class) {
+            if (xdsLocality != null) {
+                return Optional.of(xdsLocality);
+            }
+            Set<ServiceInstance> serviceInstances = serviceDiscovery
+                    .getServiceInstance(ConfigManager.getConfig(ServiceMeta.class).getService());
+            String podIp = NetworkUtils.getKubernetesPodIp();
+            if (StringUtils.isEmpty(podIp)) {
+                return Optional.empty();
+            }
+            Optional<ServiceInstance> serviceInstance = matchServiceInstanceByPodIp(serviceInstances, podIp);
+            if (serviceInstance.isPresent()) {
+                Optional<XdsLocality> validXdsLocality = createValidXdsLocality(serviceInstance.get().getMetaData());
+                xdsLocality = validXdsLocality.orElse(null);
+                return validXdsLocality;
+            }
+            return Optional.empty();
+        }
+    }
+
+    private static Optional<ServiceInstance> matchServiceInstanceByPodIp(Set<ServiceInstance> serviceInstances,
+            String podIp) {
+        return serviceInstances.stream()
+                .filter(serviceInstance -> podIp.equals(serviceInstance.getHost()))
+                .findFirst();
+    }
+
+    private static Optional<XdsLocality> createValidXdsLocality(Map<String, String> metaData) {
+        XdsLocality locality = new XdsLocality();
+        String region = metaData.get("region");
+        if (StringUtils.isEmpty(region)) {
+            return Optional.empty();
+        }
+        locality.setRegion(region);
+        String zone = metaData.get("zone");
+        String subZone = metaData.get("sub_zone");
+        if (StringUtils.isEmpty(zone) && !StringUtils.isEmpty(subZone)) {
+            return Optional.empty();
+        }
+        locality.setZone(zone);
+        locality.setSubZone(subZone);
+        return Optional.of(locality);
```

Comment:
The conditions of "return Optional.empty()" can be placed together.

Reply:
done